### PR TITLE
Support user defined CMAKE_INSTALL_INCLUDEDIR values

### LINKED ISSUE #1541: Allow Thrust and CUB headers to be nested in a custom directory
RAPIDS recently ran into a bug where the Thrust headers from a custom install were hidden by the Thrust headers in CTK, despite following the usual advice of including non-CTK Thrust headers with `-I` to override the implicit `-isystem <CTK_INCLUDE_PATH>`. 

This is due to the custom include prefix being used by multiple projects, some of which are `-isystem`, and apparently compilers will use `-isystem` when the same path is specified with both `-I` and `-isystem`. A more detailed description of the issue is in rapidsai/rapids-cmake#98.

Currently, rapids-cmake is working around this by reproducing a copy our install rules, adjusted so that thrust/cub headers are installed to `<prefix>/include/thrust/thrust` and `<prefix>/include/cub/cub`.

We should provide a more robust solution, since it's likely that other users will also encounter this issue, and I don't want to break rapids if our install rules change. We should add some CMake variables to specify this extra directory, something like `THRUST_INSTALL_HEADER_INFIX=thrust` and `CUB_INSTALL_HEADER_INFIX=cub`, or something similar (I'm not sure if there's a more idiomatic way to say "infix" in CMake).

### PR DESCRIPTION
**Pull request by robertmaynard**
Fixes #1541 

Requires CUB PR: https://github.com/NVIDIA/cub/pull/415

### CODE REVIEW

**Comment by robertmaynard** [comment]
Thinking about this more I believe we can make this easier by removing the custom INFIX variable and instead allow the user to specify a custom `CMAKE_INSTALL_INCLUDEDIR` and propagate that during configuration of the installed viles

**Comment by robertmaynard** [comment]
> Thinking about this more I believe we can make this easier by removing the custom INFIX variable and instead allow the user to specify a custom `CMAKE_INSTALL_INCLUDEDIR` and propagate that during configuration of the installed files

Updated the logic to install based on `CMAKE_INSTALL_INCLUDEDIR`. Not only did this make the logic easier to follow, it matches CMake best practises.

File: cmake/ThrustInstallRules.cmake [hunk 1]
```diff
@@ -5,15 +5,22 @@ include(GNUInstallDirs)
 set(CMAKE_SKIP_INSTALL_ALL_DEPENDENCY TRUE)
 
 install(DIRECTORY "${Thrust_SOURCE_DIR}/thrust"
-  TYPE INCLUDE
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
```
Comment (allisonvacanti) [inline cmake/ThrustInstallRules.cmake:8]: Add quotes around path.

File: cmake/ThrustInstallRules.cmake [hunk 2]
```diff
@@ -5,15 +5,22 @@ include(GNUInstallDirs)
...
   FILES_MATCHING
     PATTERN "*.h"
     PATTERN "*.inl"
 )
 
+# Need to configure a file to store THRUST_INSTALL_HEADER_INFIX
```
Comment (allisonvacanti) [inline cmake/ThrustInstallRules.cmake:14]: Comment is outdated

File: cmake/ThrustInstallRules.cmake [hunk 3]
```diff
@@ -22,12 +29,19 @@ install(DIRECTORY "${Thrust_SOURCE_DIR}/thrust/cmake/"
 option(THRUST_INSTALL_CUB_HEADERS "Include cub headers when installing." ON)
 if (THRUST_INSTALL_CUB_HEADERS)
   install(DIRECTORY "${Thrust_SOURCE_DIR}/dependencies/cub/cub"
-    TYPE INCLUDE
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
```
Comment (allisonvacanti) [inline cmake/ThrustInstallRules.cmake:32]: Quotes

File: thrust/cmake/thrust-config-version.cmake [hunk 4]
```diff
@@ -1,12 +1,6 @@
 # Parse version information from version.h:
-unset(_THRUST_VERSION_INCLUDE_DIR CACHE) # Clear old result to force search
-find_path(_THRUST_VERSION_INCLUDE_DIR thrust/version.h
-  NO_DEFAULT_PATH # Only search explicit paths below:
-  PATHS
-    ${CMAKE_CURRENT_LIST_DIR}/../..            # Source tree
-    ${CMAKE_CURRENT_LIST_DIR}/../../../include # Install tree
-)
-set_property(CACHE _THRUST_VERSION_INCLUDE_DIR PROPERTY TYPE INTERNAL)
+include(${CMAKE_CURRENT_LIST_DIR}/thrust-header-search.cmake)
```
Comment (allisonvacanti) [inline thrust/cmake/thrust-config-version.cmake:2]: Quotes around path

File: thrust/cmake/thrust-header-search.cmake [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Parse version information from version.h:
+unset(_THRUST_VERSION_INCLUDE_DIR CACHE) # Clear old result to force search
+find_path(_THRUST_VERSION_INCLUDE_DIR thrust/version.h
+  NO_DEFAULT_PATH # Only search explicit paths below:
+  PATHS
+    ${CMAKE_CURRENT_LIST_DIR}/../..            # Source tree
+)
+set_property(CACHE _THRUST_VERSION_INCLUDE_DIR PROPERTY TYPE INTERNAL)
```
Comment (allisonvacanti) [inline thrust/cmake/thrust-header-search.cmake:8]: Missing newline at EOF

File: thrust/cmake/thrust-header-search.cmake [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Parse version information from version.h:
+unset(_THRUST_VERSION_INCLUDE_DIR CACHE) # Clear old result to force search
+find_path(_THRUST_VERSION_INCLUDE_DIR thrust/version.h
+  NO_DEFAULT_PATH # Only search explicit paths below:
+  PATHS
+    ${CMAKE_CURRENT_LIST_DIR}/../..            # Source tree
```
Comment (allisonvacanti) [inline thrust/cmake/thrust-header-search.cmake:6]: Quotes

File: thrust/cmake/thrust-header-search.cmake.in [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Parse version information from version.h:
+unset(_THRUST_VERSION_INCLUDE_DIR CACHE) # Clear old result to force search
+find_path(_THRUST_VERSION_INCLUDE_DIR thrust/version.h
+  NO_DEFAULT_PATH # Only search explicit paths below:
+  PATHS
+    ${CMAKE_CURRENT_LIST_DIR}/../../../@CMAKE_INSTALL_INCLUDEDIR@
```
Comment (allisonvacanti) [inline thrust/cmake/thrust-header-search.cmake.in:6]: Quotes

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

**Comment by allisonvacanti** [comment]
LGTM

run tests

**Review by allisonvacanti — APPROVED** [review]

**Comment by allisonvacanti** [comment]
Last pushes just rebased and bumped the CUB submodule to resolve conflicts.

**Pull request merged**